sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2027,65 +1952,76 @@ class Analyzer(override val catalogManager: CatalogManager)
-            withPosition(u) {
-              Alias(GroupingID(Nil), VirtualColumn.hiveGroupingIdName)()
-            }
-          case u @ UnresolvedGenerator(name, children) =>
-            withPosition(u) {
-              v1SessionCatalog.lookupFunction(name, children) match {
-                case generator: Generator => generator
-                case other =>
-                  failAnalysis(s"$name is expected to be a generator. However, " +
-                    s"its class is ${other.getClass.getCanonicalName}, which is not a generator.")
-              }
-            }
-          case u @ UnresolvedFunction(funcId, arguments, isDistinct, filter) =>
-            withPosition(u) {
-              v1SessionCatalog.lookupFunction(funcId, arguments) match {
-                // AggregateWindowFunctions are AggregateFunctions that can only be evaluated within
-                // the context of a Window clause. They do not need to be wrapped in an
-                // AggregateExpression.
-                case wf: AggregateWindowFunction =>
-                  if (isDistinct || filter.isDefined) {
-                    failAnalysis("DISTINCT or FILTER specified, " +
-                      s"but ${wf.prettyName} is not an aggregate function")
-                  } else {
-                    wf
-                  }
-                // We get an aggregate function, we need to wrap it in an AggregateExpression.
-                case agg: AggregateFunction =>
-                  if (filter.isDefined && !filter.get.deterministic) {
-                    failAnalysis("FILTER expression is non-deterministic, " +
-                      "it cannot be used in aggregate functions")
-                  }
-                  AggregateExpression(agg, Complete, isDistinct, filter)
-                // This function is not an aggregate function, just return the resolved one.
-                case other if (isDistinct || filter.isDefined) =>
-                  failAnalysis("DISTINCT or FILTER specified, " +
-                    s"but ${other.prettyName} is not an aggregate function")
-                case e: String2TrimExpression if arguments.size == 2 =>
-                  if (trimWarningEnabled.get) {
-                    log.warn("Two-parameter TRIM/LTRIM/RTRIM function signatures are deprecated." +
-                      " Use SQL syntax `TRIM((BOTH | LEADING | TRAILING)? trimStr FROM str)`" +
-                      " instead.")
-                    trimWarningEnabled.set(false)
-                  }
-                  e
-                case other =>
-                  other
-              }
-            }
+      case a: Aggregate =>
+        val newGroups = a.groupingExpressions.map {
+          case gs: GroupingSet =>
+            gs.withNewChildren(gs.children.map(_.transformDown(resolveFunction)))
+          case e => e
         }
+        a.copy(groupingExpressions = newGroups) transformExpressions resolveFunction
```

Comment:
We still need this entry?

Reply:
> We still need this entry?

Yea，still need this logical. Tried to merge code, but failed.

Reply:
IMO these entries is still annoying... For example, these entries do not seem to be needed in my suggested approach above: https://github.com/maropu/spark/commit/2da2da6812f3119715e46e2acbc3faa2e314fd81 So, could you think more about another approach for removing them?

Reply:
> IMO these entries is still annoying... For example, these entries do not seem to be needed in my suggested approach above: [maropu@2da2da6](https://github.com/maropu/spark/commit/2da2da6812f3119715e46e2acbc3faa2e314fd81) So, could you think more about another approach for removing them?

How about current? It can work well.
```
      case q: LogicalPlan => q transformExpressions {
        case gs: GroupingSet =>
          gs.withNewChildren(gs.children.map(_.transformDown { case e => resolveFunction(e) }))
        case e => resolveFunction(e)
      }
```

Reply:
I think its better not to make the existing Analyzer logics complicated where possible. Handling all the case in ` case e => resolveFunction(e)` looks the best.

Reply:
> I think its better not to make the existing Analyzer logics complicated where possible. Handling all the case in ` case e => resolveFunction(e)` looks the best.

How about current? I have no ideal about how to integrate `resolveExpr` and `reslveFunction`

Reply:
Actually, I'm not sure now about why we need to update `ResolveFunctions`. If `GroupingSet`(`Rollup`, `Cube`, and `GroupingSet`) has a correct structure like the other expressions, I don't think we need to update this rule at all. If it is unavoidable to update it, I think we need a reasonable reason to do so.

Reply:
> Actually, I'm not sure now about why we need to update `ResolveFunctions`. If `GroupingSet`(`Rollup`, `Cube`, and `GroupingSet`) has a correct structure like the other expressions, I don't think we need to update this rule at all. If it is unavoidable to update it, I think we need a reasonable reason to do so.

emmm, without this change some test failed, later I will debug to find the real reason.

Reply:
@maropu  Sorry for late reply..Now we can handle all in `resolveFucntion` like you mentioned  
 in https://github.com/apache/spark/pull/30212#discussion_r533090836
But I don't know why it won't resolve `GroupingSet`'s children..

Reply:
It's strange that if I call `mapChildren` of `GroupingSet`, it won't loop the children.
But call `children` it can have value. And if I call `expr.containsChild` before 'mapChildren',  it can work well.
Any suggestion? it's so confused.